### PR TITLE
Handle no peer certificates cleanly

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
https://github.com/square/okhttp/issues/4208

File: okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java [hunk 1]
```diff
@@ -320,9 +321,16 @@ private void connectTls(ConnectionSpecSelector connectionSpecSelector) throws IO
...
       Handshake unverifiedHandshake = Handshake.get(sslSocketSession);
 
+      List<Certificate> peerCertificates = unverifiedHandshake.peerCertificates();
+
       // Verify that the socket's certificates are acceptable for the target host.
       if (!address.hostnameVerifier().verify(address.url().host(), sslSocketSession)) {
-        X509Certificate cert = (X509Certificate) unverifiedHandshake.peerCertificates().get(0);
+        if (peerCertificates.isEmpty()) {
+          throw new SSLPeerUnverifiedException(
+              "Hostname " + address.url().host() + " not verified, no certificates provided");
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: Apparently there's a thing called “pre-shared keys” that make this case valid! Not sure how to check that though.

Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java]: Valid as in we should check if that is the case and allow those connections?  I'll look into it.

**Comment by francaracuel** [comment]
Hi @yschimke, we got the same crash that appears here: https://github.com/square/okhttp/issues/4208.
Do you know if, finally, are you going to merge it to the master branch?
Thank you!

**Comment by yschimke** [comment]
@francaracuel No, the proper fix was PSK support, and I haven't had a chance to test.

@swankjesse is there a clear handling that would fix this bug for now?

**Comment by swankjesse** [comment]
I think we need a PSK test.

PSK won't have a regular handshake, and so if we support this exotic feature then **everyone** will need to nullcheck their handshakes even on HTTPS connections. Previously we could conveniently say all HTTPS calls had handshakes.

Kind of makes me want to figure out a way to make PSK look like something with a handshake.

**Comment by yschimke** [comment]
I'd rather not fake the test based on how I understand PSK to work.  But conscrypt deprecated it because it doesn't fit with TLS 1.3 apparently, and I haven't found a nice clear sample of configuring it.

**Comment by yuforia** [comment]
Please fix soon,  this has become a really big problem.
@swankjesse We are not using anything special like PSK, this just happens irregularly.
Is there a way to workaround the problem? How can the out of bound exception be caught and the operation of my program resumed?
Currently the "OkHttp Dispatcher" thread crashes and I don't know how to catch the exception and have to restart the whole program.

**Comment by swankjesse** [comment]
@yuforia can you create a new issue with a test case that reproduces this?

**Pull request closed without merging**